internal/controller/dnsrecord_controller.go

```diff
@@ -407,10 +418,19 @@ func (r *DNSRecordReconciler) applyChanges(ctx context.Context, dnsRecord *v1alp
 		}
 	} else {
 		logger.Info("All records are already up to date")
+		// reset the valid for from randomized value to a fixed value once validation succeeds
+		if !meta.IsStatusConditionTrue(dnsRecord.Status.Conditions, string(v1alpha1.ConditionTypeReady)) {
+			requeueTime = exponentialRequeueTime(validationRequeueTime.String())
+			dnsRecord.Status.ValidFor = requeueTime.String()
+		}
+		setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionTrue, "ProviderSuccess", "Provider ensured the dns record")
+	}
+
+	// reset the counter on the gen change regardless of having changes in the plan
+	if generationChanged(dnsRecord) {
 		dnsRecord.Status.WriteCounter = 0
 		wrtiteCounter.WithLabelValues(dnsRecord.Name, dnsRecord.Namespace).Set(0)
-		setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionTrue, "ProviderSuccess", "Provider ensured the dns record")
 	}
 
-	return defaultRequeueTime, nil
+	return requeueTime, nil
 }
```

Comment:
Do we have an issue to remove all this logic from applyChanges? It's really hard to read and the status is still being updated in places it shouldn't i.e. before it's even done anything

Reply:
There was one but it seems like I can't find it. Changed anyway. Now it should be easier to read\maintain